### ISSUE TITLE
overzZzealous: Non-lazy Semantics of || and && in Plutus Core

### ISSUE THREAD

**Issue by Niols** [IntersectMBO/plutus#4114]
Hi there,

While developping tools with [Hachi][gh:hachi], we noticed something that we believe is a bug in the compiler from Plutus to Plutus Core.

Best,  
Nicolas

EDIT: It seems to not be considered a bug. Since we believe it will still come up regularly in issues here and there, we decided to give this peculiarity the sweet name of “overzZzealous”. See also [our blog post on the topic][blogpost].

## Area

- [x] *Plutus Foundation*
- [ ] *Plutus Application Framework*
- [ ] *Marlowe*
- [ ] *Other*

## Summary

The compiler from Plutus to Plutus Core seems to compile `||` and `&&` as non-lazy, that is where the right-hand side is evaluated regardless of what the left-hand side evaluates to.

## Steps to reproduce

In general:

1. Take any Plutus code that runs on-chain.
2. Introduce side-effects on the right of a `||` or `&&`. (eg. `if True || error () then do_something else True`)
3. Observe that the side-effect is triggered.

More guided, within the *Plutus Application Backend (PAB)* example:

1. Clone the [input-output-hk/plutus-starter][gh:iohk/ppsp] project.
2. Follow the README to setup the project and to play the game with the PAB. Check that everything works: you get a validation error if you guess wrong, and the transation is validated if you guess right.
3. Update `validateGuess` in `./examples/src/Plutus/Contracts/Game.hs` to introduce an `error ()` side-effect on the right of an `||`. We suggest to use: `validateGuess hs cs _ = if True || error () then isGoodGuess hs cs else True`.
4. Play the game again and observe that you get a validation error whether you guess right or wrong.

Check [our fork of the Plutus Platform starter project][gh:hachi/ppsp] for a more guided version, including Shell scripts to play and modify the code in your place.

## Expected behavior

We would expect `||` and `&&` to be lazy and to only evaluate their second argument only if the first one evaluates to `False` and `True` respectively. This expectation is based on the fact that it is the case in a lot of languages and therefore we believe programmers will expect this as well. Moreover, it seems to be the case both [in Haskell][haskell-or] and [in off-chain Plutus][plutus-or].

## System info

Tested on several machines, all with latest Plutus.

- 1:
  - OS: NixOS 21.05, up-to-date as of Friday, October 15th, 2021.
  - GHC: 8.10.4
- 2:
  - OS: MacOS 11.6
  - GHC: 8.10.4

## Attachments

See our blog post on the topic: [*overzZzealous*: the Peculiar Semantics of `||` and `&&` in Plutus Core][blogpost].

[blogpost]: https://blog.hachi.one/post/overzzzealous-peculiar-semantics-or-and-plutus-core

[gh:iohk/ppsp]:  https://github.com/input-output-hk/plutus-starter
[gh:hachi]:      https://github.com/HachiSecurity
[gh:hachi/ppsp]: https://github.com/HachiSecurity/overzzzealous-in-plutus-starter

[haskell-or]: https://hackage.haskell.org/package/base-4.15.0.0/docs/Prelude.html#v:-124--124-
[plutus-or]:https://github.com/input-output-hk/plutus/blob/75ccaf4/plutus-tx/src/PlutusTx/Bool.hs#L30

**Comment by michaelpj** [comment]
> Is there a reason why that is?

Booleans aren't a special type either. Booleans and their functions are compiled just like any other datatypes and functions, there's no special handling.

**Comment by kk-hainq** [comment]
We had some time to revisit the issue today and did a quick check on this `cheapCheck || expensiveChecks` example at `plutus-use-cases`:
https://github.com/HachiSecurity/plutus-use-cases/pull/1/files

This innocent-looking code might require over 1.5 times more resources than an unnatural-looking `if-then-else`.
``` haskell
-- || is surprisingly way more expensive!
isUpdateValid = (not isCurrentValueSigned) ||
    (fromMaybe False $ validateGameStatusChanges <$>
    (osmGameStatus <$> extractSigendMessage (ovSignedMessage oracleData)) <*>
    (osmGameStatus <$> extractSigendMessage outputSignedMessage))

-- if-then-else is way cheaper!
isUpdateValid = if not isCurrentValueSigned then True else
    (fromMaybe False $ validateGameStatusChanges <$>
    (osmGameStatus <$> extractSigendMessage (ovSignedMessage oracleData)) <*>
    (osmGameStatus <$> extractSigendMessage outputSignedMessage))
```

We calculate the required budget using `writeScriptsTo` as documented in the [official how-to](https://playground.plutus.iohkdev.io/doc/plutus/howtos/analysing-scripts.html).
``` haskell
writeScriptsTo
  (ScriptsConfig "." (Scripts UnappliedValidators))
  "updateOracleTrace"
  Spec.Oracle.updateOracleTrace
  def
```

The difference is rather significant:
``` haskell
-- With `||`:
( Sum {getSum = 9539}
, ExBudget {exBudgetCPU = ExCPU 1124168956, exBudgetMemory = ExMemory 3152700}
)

-- With `if-then-else`:
( Sum {getSum = 3531}
, ExBudget {exBudgetCPU = ExCPU 698345767, exBudgetMemory = ExMemory 2069978}
)
```

When the quick check is satisfied, the strict `||` still evaluates the heavy-weight right-hand side that yields a much larger script and requires over 1.5 times more resources than a short-circuit solution in `if-then-else`. The behaviour might be extremely costly and wasteful had a frequently updated oracle feed falls into this case.

The worst part is that we had no clues about this behaviour until we diligently built a PLC concolic execution engine. Otherwise, we would write such wasteful code ourselves, given how natural that code looks like. `hlint` itself suggests the change!
``` txt
src/Contracts/Oracle/OnChain.hs:(136,21)-(139,69): Warning: Redundant if
Found:
  if not isCurrentValueSigned then
      True
  else
      (fromMaybe False
         $ validateGameStatusChanges
             <$>
               (osmGameStatus
                  <$> extractSigendMessage (ovSignedMessage oracleData))
             <*> (osmGameStatus <$> extractSigendMessage outputSignedMessage))
Perhaps:
  not isCurrentValueSigned
    ||
      (fromMaybe False
         $ validateGameStatusChanges
             <$>
               (osmGameStatus
                  <$> extractSigendMessage (ovSignedMessage oracleData))
             <*> (osmGameStatus <$> extractSigendMessage outputSignedMessage))
```

We'd love to help document and communicate the behaviour to end developers. We'll add another example with dangerous side-effects, search for deployed scripts with the expensive/side-effect patterns, and implement a fix if the team decided on any.

**Comment by michaelpj** [comment]
Plutus Core is a strict language, for better or worse at this point. Yes we need to communicate that better, I agree. Possibly we just need an explainer piece in the documentation.

**Comment by kk-hainq** [comment]
We've already been writing a blog post documenting this topic in more depth. If it's welcomed I can port the content rewritten in formal documentation style to a PR somewhere appropriate.

I've talked to a few developers and they're probably going to rewrite all `||` and `&&` with `if-then-else` or `case-of` like Edmund did.

Code like quoted below would become very hard to read, but most products desire cheaper transaction fees, even faster off-chain validations (few users would run their own node, and project nodes are already accompanied with a currently-still-slow chain index and all the extra infrastructure required) at the moment. Alternatively, developers can compile with an ad-hoc fork that has an extra commit that compiles `||` and `&&` lazily. We will not convince people which is the way, just let the developers decide based on their needs.

https://github.com/input-output-hk/plutus-apps/blob/c3676cfe11bcd8512d7353ee337de8e84c7f2278/plutus-use-cases/src/Plutus/Contracts/Swap.hs#L133-L163

``` haskell
    ...
    iP1 :: TxInInfo -> Bool
    iP1 TxInInfo{txInInfoResolved=TxOut{txOutValue}} = ...

    iP2 :: TxInInfo -> Bool
    iP2 TxInInfo{txInInfoResolved=TxOut{txOutValue}} = ...

    inConditions = (iP1 t1 && iP2 t2) || (iP1 t2 && iP2 t1)

    ol1 :: TxOut -> Bool
    ol1 o@TxOut{txOutValue} = ...

    ol2 :: TxOut -> Bool
    ol2 o@TxOut{txOutValue} = ...

    outConditions = (ol1 o1 && ol2 o2) || (ol1 o2 && ol2 o1)

in inConditions && outConditions
```

In conclusion, most developers I've talked to agree that this is very nice to have, but not critical at the moment. They would love us to reduce script sizes more, for instance. Let's document it nicely and move on for now I guess.

**Comment by anton-k** [comment]
I've defined possible solution to clumsy `if-then-else` expansion with TH.
We can use TH to transform `and` and `or` expressions to boring long versions that implement short circuit.

Here is PR #4191

**Comment by kk-hainq** [comment]
@anton-k Awesome! That is definitely useful in several use cases :+1:.

**Comment by treeowl** [comment]
I may be a outlier here, but I don't think it makes any sense to compile GHC Core as though it were for a strict language. It would make much more sense to me to treat it as exactly the language it is, and to expect user-written code to be as strict as it should be. Doing this efficiently does mean changing some datatypes. For example, a builtin list read from the blockchain should be encoded as an entirely *strict* list. Eventually, it will make sense to use *unlifted* datatypes (9.2.1 doesn't handle those right for worker/wrapper; the fix might get backported for 9.2.2). But the current approach is much too wild. Perfectly reasonable core2core transformations can produce wildly different performance in Plutus, and there are all sorts of horrible hacks to try to avoid that.

**Comment by michaelpj** [comment]
@treeowl I don't think you're an outlier. I think it's pretty arguable that the approach we take isn't the best one. One alternative would be what you expect: get users to write strict Haskell if they want strict Plutus Core. But this also leaves people lots of scope for footguns in terms of making things less trict than they meant to. And, as you point out, doing it in a systematic and principled fashion would really need `UnliftedDatatypes`, which is some way away from being usable in practice, I fear. Maybe we could get away with making things generally non-strict, but I'm pretty wary.

**Comment by treeowl** [comment]
@michaelpj, yes, there would be footguns, but I think a lot fewer than there are now. This is just too wild to be able to program. Have you been able to take a look at some of my recent questions in #4206? I could use a bit of help.

**Comment by aquynh** [comment]
Hachi team had a detailed writeup on this issue at https://blog.hachi.one/post/overzzzealous-peculiar-semantics-or-and-plutus-core. Hopefully, this raises more awareness in the Cardano community, helping developers to avoid the problem and write more efficient validator scripts.

**Comment by Mercurial** [comment]
> **Hachi** team had a detailed writeup on this issue at https://blog.hachi.one/post/overzzzealous-peculiar-semantics-or-and-plutus-core.
> 
> Hopefully, this raises awareness in the Cardano community, helping developers to avoid the issue and write more efficient validator scripts.

this definitely raised awareness and there were some unexpected behaviors with our && operator inside a validator. We'll report back with more information once we apply the knowledge shared in the article if it was really a problem of Plutus Core or was it on our side.

**Comment by Niols** [comment]
Since it is not an issue but the correct semantics, I suppose we can close this issue?

**Comment by effectfully** [comment]
I'm annoyed by the problem and want to keep the issue open until we come to the conclusion that we don't care.

**Comment by L-as** [comment]
Fixed by https://github.com/cardano-foundation/CIPs/pull/469

**Comment by L-as** [comment]
The above CIP in fact also fixes the issue with `expensive1 || expensive2` being suboptimal.

**Comment by L-as** [comment]
You can just have a RULE for it @effectfully
